cql3/cql3_type.cc

```diff
@@ -303,6 +303,47 @@ class cql3_type::raw_tuple : public raw {
     }
 };
 
+class cql3_type::raw_vector : public raw {
+    shared_ptr<raw> _type;
+    size_t _dimension;
+
+    virtual sstring to_string() const override {
+        return seastar::format("vector<{}, {}>", _type, _dimension);
+    }
+
+public:
+    raw_vector(shared_ptr<raw> type, size_t dimension)
+            : _type(std::move(type)), _dimension(dimension) {
+    }
+
+    virtual bool supports_freezing() const override {
+        return true;
+    }
+
+    virtual bool is_collection() const override {
+        return false;
+    }
+
+    virtual void freeze() override {
+        if (_type->supports_freezing()) {
+            _type->freeze();
+        }
+        _frozen = true;
+    }
```

Comment:
Looks like freezing a vector does not have any effect:

```
cqlsh> create table ks.t (a int, b frozen<vector<list<int>, 32>>, c vector<frozen<list<int>>, 16>, primary key (a));
cqlsh> describe ks.t;

CREATE TABLE ks.t (
    a int PRIMARY KEY,
    b vector<list<int>, 32>,
    c vector<frozen<list<int>>, 16>
) WITH additional_write_policy = '99p'
    AND allow_auto_snapshot = true
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND cdc = false
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '16', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND memtable = 'default'
    AND crc_check_chance = 1.0
    AND default_time_to_live = 0
    AND extensions = {}
    AND gc_grace_seconds = 864000
    AND incremental_backups = true
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair = 'BLOCKING'
    AND speculative_retry = '99p';
```

What's interesting, it doesn't seem to freeze the inner type. I am not sure whether this was intended in the original implementation, but let's be compatible.

Reply:
To match Apache Cassandra's implementation, freezing of inner elements has been removed.